plugins/woocommerce/includes/wc-core-functions.php

```diff
@@ -1306,6 +1306,50 @@ function wc_get_base_location() {
 	return wc_format_country_state_string( $default );
 }
 
+/**
+ * Uses geolocation to get the customer country and state only if they are valid values.
+ *
+ * @since 9.5.0
+ * @param array $fallback Fallback location.
+ * @return array
+ */
+function wc_get_customer_geolocation( $fallback = array(
+	'country' => '',
+	'state'   => '',
+) ) {
+	$ua = wc_get_user_agent();
+
+	// Exclude common bots from geolocation by user agent.
+	if ( stristr( $ua, 'bot' ) || stristr( $ua, 'spider' ) || stristr( $ua, 'crawl' ) ) {
+		return $fallback;
+	}
+
+	$geolocation = WC_Geolocation::geolocate_ip( '', true, false );
+
+	if ( empty( $geolocation['country'] ) ) {
+		return $fallback;
+	}
+
+	// Ensure geolocation is valid.
+	$allowed_countries = WC()->countries->get_allowed_countries();
+
+	if ( ! in_array( $geolocation['country'], array_keys( $allowed_countries ), true ) ) {
```

Comment:
Something made me feel weird reading this line... since arrays keys are case sensitive, isn't this the same as:
`if ( ! isset( $allowed_countries[ $geolocation['country'] ] ) ) {...}`
(But without the overhead of creating another array through `array_keys()`)

Reply:
GPT agrees